exporter/exporterhelper/tracehelper.go

```diff
@@ -86,57 +83,91 @@ func (p traceDataPusherOld) withObservability(exporterName string) traceDataPush
 	}
 }
 
+// traceDataPusher is a helper function that is similar to ConsumeTraceData but also
+// returns the number of dropped spans.
+type traceDataPusher func(ctx context.Context, td pdata.Traces) (droppedSpans int, err error)
+
+type tracesRequest struct {
+	baseRequest
+	td     pdata.Traces
+	pusher traceDataPusher
+}
+
+func newTracesRequest(ctx context.Context, td pdata.Traces, pusher traceDataPusher) request {
+	return &tracesRequest{
+		baseRequest: baseRequest{ctx: ctx},
+		td:          td,
+		pusher:      pusher,
+	}
+}
+
+func (req *tracesRequest) onPartialError(partialErr consumererror.PartialError) request {
+	return newTracesRequest(req.ctx, partialErr.GetTraces(), req.pusher)
+}
+
+func (req *tracesRequest) export(ctx context.Context) (int, error) {
+	return req.pusher(ctx, req.td)
+}
+
+func (req *tracesRequest) count() int {
+	return req.td.SpanCount()
+}
+
 type traceExporter struct {
-	baseExporter
-	dataPusher traceDataPusher
+	*baseExporter
+	pusher traceDataPusher
 }
 
-func (te *traceExporter) ConsumeTraces(
-	ctx context.Context,
-	td pdata.Traces,
-) error {
-	exporterCtx := obsreport.ExporterContext(ctx, te.exporterFullName)
-	_, err := te.dataPusher(exporterCtx, td)
+func (texp *traceExporter) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
+	exporterCtx := obsreport.ExporterContext(ctx, texp.cfg.Name())
+	req := newTracesRequest(exporterCtx, td, texp.pusher)
+	_, err := texp.sender.send(req)
 	return err
 }
 
 // NewTraceExporter creates a TraceExporter that can record metrics and can wrap
 // every request with a Span.
 func NewTraceExporter(
-	config configmodels.Exporter,
+	cfg configmodels.Exporter,
 	dataPusher traceDataPusher,
 	options ...ExporterOption,
 ) (component.TraceExporter, error) {
 
-	if config == nil {
+	if cfg == nil {
 		return nil, errNilConfig
 	}
 
 	if dataPusher == nil {
 		return nil, errNilPushTraceData
 	}
 
-	dataPusher = dataPusher.withObservability(config.Name())
+	be := newBaseExporter(cfg, options...)
+
+	// Record metrics on the consumer.
+	be.qSender.nextSender = &tracesExporterWithObservability{
+		exporterName: cfg.Name(),
+		sender:       be.qSender.nextSender,
```

Comment:
Same point about `qSender` vs. `sender`.

Reply:
Same comment about observability that needs to know about producer and consumer.